codefeedr-plugins/codefeedr-travis/src/main/scala/org/codefeedr/plugins/travis/util/TravisBuildCollector.scala

```diff
@@ -0,0 +1,158 @@
+  * @param repoOwner Owner of the repo of the push event
+  * @param repoName Name of the repo of the push event
+  * @param branchName Name of the branch of the push event
+  * @param pushCommitSha Sha of the commit of the push event
+  * @param pushDate Date of the push event
+  * @param travis TravisService used to query Travis
+  * @param pollingInterval Interval at which builds are queried
+  * @param timeoutSeconds How long it will wait for the build to be created
+  */
+class TravisBuildCollector(repoOwner: String,
+                           repoName: String,
+                           branchName: String,
+                           pushCommitSha: String,
+                           pushDate: LocalDateTime,
+                           travis: TravisService,
+                           pollingInterval: Int = 30000,
+                           timeoutSeconds: Int = 60) {
+
+  private var minimumStartDate: LocalDateTime = pushDate
+  private var build: Option[TravisBuild] = None
+
+
+  /**
+    * Keeps requesting the build until it is finished
+    * @return A finished Travis build
+    */
+  def requestFinishedBuild(): Future[TravisBuild] = Future {
+    while (!isReady) {
+
+      build = requestBuild
+      checkIfBuildShouldBeKnownAlready()
+
+      if (!isReady) {
+        blocking {
+          Thread.sleep(pollingInterval)
+        }
+      }
+    }
+    build.get
+  }
+
+  def checkIfBuildShouldBeKnownAlready(): Unit = {
+    val waitUntil = pushDate.plusSeconds(100 * timeoutSeconds)
+    if (build.isEmpty && waitUntil.isBefore(LocalDateTime.now(ZoneId.of("GMT")))) {
+      throw BuildNotFoundForTooLongException("Waited " + timeoutSeconds + " seconds for build, but still not found" +
+        ", probably because " + repoOwner + "/" + repoName + "is not active on Travis")
+    }
+  }
+
+  /**
+    * Checks whether or not a build is completed.
+    * @return Whether or not a build is completed
+    */
+  def isReady: Boolean = {
+    if (build.nonEmpty) {
+      val state = build.get.state
+      return state == "passed" || state == "failed" || state == "canceled" || state == "errored"
+    }
+    false
+  }
+
+  /**
+    * Requests the build information based on if a build id is known
+    * @return A Travis build
+    */
+  def requestBuild: Option[TravisBuild] = {
+    try {
+      build match {
+        case None => requestUnknownBuild()
+        case Some(_) => requestKnownBuild()
+      }
+    } catch {
+      case _: CouldNotExtractException =>
+        throw CouldNotAccessTravisBuildInfo("Could not retrieve Travis build info for: "
+          + repoOwner + "/" + repoName)
+      case _: CouldNotGetResourceException =>
+        None
+    }
+  }
+
+  /**
+    * Requests a build from travis of which the build id is known
+    * @return A TravisBuild if it is found, None otherwise
+    */
+  def requestKnownBuild(): Option[TravisBuild] = {
+    assert(build.nonEmpty)
```

Comment:
How is this assert used? Will it throw an exception if it is not true?

Reply:
This will throw an exception if the build is indeed empty. The flink job will probably crash as well, but that's okay since there is something wrong with the code if this happens